File: boat-core/src/test/java/test/java/xyz/srclab/common/base/RefTest.java
package test.java.xyz.srclab.common.base;

import org.testng.Assert;
import org.testng.annotations.Test;
import xyz.srclab.common.base.Ref;
import xyz.srclab.common.logging.Logger;

import java.time.LocalDateTime;
import java.time.ZoneId;
import java.time.chrono.ChronoZonedDateTime;
import java.time.format.DateTimeFormatter;
import java.util.Date;

public class RefTest {

    private static final Logger logger = Logger.simpleLogger();

    @Test
    public void testRef() {
        Ref<String> ref = Ref.of(null);
        Assert.assertEquals(ref.getOrElse("null"), "null");
        Assert.assertFalse(ref.isPresent());
        ref.set("123");
        Assert.assertEquals(ref.getOrElse("null"), "123");
        Assert.assertTrue(ref.isPresent());
    }

    @Test
    public void testChainOps() {
        String dateString = "2222-12-22 22:22:22";
        DateTimeFormatter formatter = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss");
        Date date = Ref.of(dateString)
            .accept(logger::info)
            .apply(d -> LocalDateTime.parse(d, formatter))
            .apply(d -> d.atZone(ZoneId.systemDefault()))
            .with(ChronoZonedDateTime::toInstant)
            .with(Date::from)
            .get();
        logger.info("date: {}", date);
    }
}

File: easy-common/src/test/kotlin/test/xyz/srclab/common/lang/RefTest.kt
package test.xyz.srclab.common.lang

import org.testng.annotations.Test
import test.xyz.srclab.common.doAssertEquals
import xyz.srclab.common.lang.Ref

/**
 * @author sunqian
 */
object RefTest {

    @Test
    fun testRef() {
        val ref = Ref.withEmpty<Int>()
        doAssertEquals(ref.get(), null)
        ref.set(1)
        doAssertEquals(ref.get(), 1)

        val ref2 = Ref.with("1")
        doAssertEquals(ref.get(), 1)
    }
}